File: zig/dsa/invert-binary-tree/invert_binary_tree.zig
// Invert a binary tree (mirror left/right).
//
//        4                 4
//      /   \             /   \
//     2     7    -->    7     2
//    / \   / \         / \   / \
//   1   3 6   9       9   6 3   1
//
// Build/run:  zig run invert_binary_tree.zig

const std = @import("std");

const Node = struct {
    val: i32,
    left: ?*Node = null,
    right: ?*Node = null,
};

fn newNode(alloc: std.mem.Allocator, val: i32, left: ?*Node, right: ?*Node) !*Node {
    const n = try alloc.create(Node);
    n.* = .{ .val = val, .left = left, .right = right };
    return n;
}

fn invert(root: ?*Node) void {
    const r = root orelse return;
    const tmp = r.left;
    r.left = r.right;
    r.right = tmp;
    invert(r.left);
    invert(r.right);
}

fn levelOrder(root: ?*Node, out: []i32) usize {
    const r = root orelse return 0;
    var queue: [32]*Node = undefined;
    var head: usize = 0;
    var tail: usize = 0;
    queue[tail] = r;
    tail += 1;
    var count: usize = 0;
    while (head < tail) {
        const n = queue[head];
        head += 1;
        out[count] = n.val;
        count += 1;
        if (n.left) |l| {
            queue[tail] = l;
            tail += 1;
        }
        if (n.right) |rc| {
            queue[tail] = rc;
            tail += 1;
        }
    }
    return count;
}

fn printInts(label: []const u8, xs: []const i32) void {
    std.debug.print("{s}", .{label});
    for (xs) |x| std.debug.print("{d} ", .{x});
    std.debug.print("\n", .{});
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const root = try newNode(a, 4,
        try newNode(a, 2,
            try newNode(a, 1, null, null),
            try newNode(a, 3, null, null)),
        try newNode(a, 7,
            try newNode(a, 6, null, null),
            try newNode(a, 9, null, null)));

    var buf: [16]i32 = undefined;

    var n = levelOrder(root, &buf);
    printInts("Before: ", buf[0..n]);

    invert(root);

    n = levelOrder(root, &buf);
    printInts("After:  ", buf[0..n]);

    const expected = [_]i32{ 4, 7, 2, 9, 6, 3, 1 };
    std.debug.assert(n == expected.len);
    for (buf[0..n], expected) |got, exp| std.debug.assert(got == exp);
    std.debug.print("OK — inverted tree matches expected\n", .{});
}

File: c/dsa/invert-binary-tree/invert_binary_tree.c
// Invert a binary tree (mirror left/right).
/*
        4                 4
      /   \             /   \
     2     7    -->    7     2
    / \   / \         / \   / \
   1   3 6   9       9   6 3   1
*/
// Build:   cc -O2 -Wall -Wextra -o invert invert_binary_tree.c
// Run:     ./invert

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int val;
    struct Node *left, *right;
} Node;

static Node *new_node(int val, Node *left, Node *right) {
    Node *n = malloc(sizeof(*n));
    n->val = val;
    n->left = left;
    n->right = right;
    return n;
}

static void invert(Node *root) {
    if (!root) return;
    Node *tmp = root->left;
    root->left = root->right;
    root->right = tmp;
    invert(root->left);
    invert(root->right);
}

static int level_order(Node *root, int *out, int cap) {
    if (!root) return 0;
    Node *queue[64];
    int head = 0, tail = 0, count = 0;
    queue[tail++] = root;
    while (head < tail && count < cap) {
        Node *n = queue[head++];
        out[count++] = n->val;
        if (n->left)  queue[tail++] = n->left;
        if (n->right) queue[tail++] = n->right;
    }
    return count;
}

static void print_ints(const char *label, const int *xs, int n) {
    printf("%s", label);
    for (int i = 0; i < n; i++) printf("%d ", xs[i]);
    printf("\n");
}

static void free_tree(Node *root) {
    if (!root) return;
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

int main(void) {
    Node *root = new_node(4,
        new_node(2, new_node(1, NULL, NULL), new_node(3, NULL, NULL)),
        new_node(7, new_node(6, NULL, NULL), new_node(9, NULL, NULL)));

    int buf[16];
    int n = level_order(root, buf, 16);
    print_ints("Before: ", buf, n);

    invert(root);

    n = level_order(root, buf, 16);
    print_ints("After:  ", buf, n);

    const int expected[] = {4, 7, 2, 9, 6, 3, 1};
    assert(n == (int)(sizeof(expected) / sizeof(*expected)));
    for (int i = 0; i < n; i++) assert(buf[i] == expected[i]);
    printf("OK — inverted tree matches expected\n");

    free_tree(root);
    return 0;
}
